node/core/provisioner/src/disputes/prioritized_selection/mod.rs

```diff
@@ -0,0 +1,460 @@
+			);
+			0
+		},
+	}
+}
+
+fn concluded_onchain(onchain_state: &DisputeState) -> bool {
+	// Check if there are enough onchain votes for or against to conclude the dispute
+	let supermajority = supermajority_threshold(onchain_state.validators_for.len());
+
+	onchain_state.validators_for.count_ones() >= supermajority ||
+		onchain_state.validators_against.count_ones() >= supermajority
+}
+
+fn partition_recent_disputes(
+	recent: Vec<(SessionIndex, CandidateHash, DisputeStatus)>,
+	onchain: &HashMap<(SessionIndex, CandidateHash), DisputeState>,
+) -> PartitionedDisputes {
+	let mut partitioned = PartitionedDisputes::new();
+
+	// Drop any duplicates
+	let unique_recent = recent
+		.into_iter()
+		.map(|(session_index, candidate_hash, dispute_state)| {
+			((session_index, candidate_hash), dispute_state)
+		})
+		.collect::<HashMap<_, _>>();
+
+	// Split recent disputes in ACTIVE and INACTIVE
+	let time_now = &secs_since_epoch();
+	let (active, inactive): (
+		Vec<(SessionIndex, CandidateHash, DisputeStatus)>,
+		Vec<(SessionIndex, CandidateHash, DisputeStatus)>,
+	) = unique_recent
+		.into_iter()
+		.map(|((session_index, candidate_hash), dispute_state)| {
+			(session_index, candidate_hash, dispute_state)
+		})
+		.partition(|(_, _, status)| !dispute_is_inactive(status, time_now));
+
+	// Split ACTIVE in three groups...
+	for (session_index, candidate_hash, _) in active {
+		match onchain.get(&(session_index, candidate_hash)) {
+			Some(d) => match concluded_onchain(d) {
+				true => partitioned.active_concluded_onchain.push((session_index, candidate_hash)),
+				false =>
+					partitioned.active_unconcluded_onchain.push((session_index, candidate_hash)),
+			},
+			None => partitioned.active_unknown_onchain.push((session_index, candidate_hash)),
+		};
+	}
+
+	// ... and INACTIVE in three more
+	for (session_index, candidate_hash, _) in inactive {
+		match onchain.get(&(session_index, candidate_hash)) {
+			Some(onchain_state) =>
+				if concluded_onchain(onchain_state) {
+					partitioned.inactive_concluded_onchain.push((session_index, candidate_hash));
+				} else {
+					partitioned.inactive_unconcluded_onchain.push((session_index, candidate_hash));
+				},
+			None => partitioned.inactive_unknown_onchain.push((session_index, candidate_hash)),
+		}
+	}
+
+	partitioned
+}
+
+// Helper trait to obtain the value of vote for `InvalidDisputeStatementKind` and `ValidDisputeStatementKind`.
+// The alternative was to pass a bool to `fn is_vote_worth_to_keep` explicitly but it's pointless as the value is already 'encoded' in the type.
+trait VoteType {
+	fn is_valid() -> bool;
+}
+
+impl VoteType for InvalidDisputeStatementKind {
+	fn is_valid() -> bool {
+		false
+	}
+}
+
+impl VoteType for ValidDisputeStatementKind {
+	fn is_valid() -> bool {
+		true
+	}
+}
+
+/// Determines if a vote is worth to be kept, based on the onchain disputes
+fn is_vote_worth_to_keep<T: VoteType>(
+	validator_index: &ValidatorIndex,
+	_: &T,
```

Comment:
For simplicity I would have just passed a bool here rather than use `VoteType`.

Reply:
I agree it's simplier.. but if you don't mind I prefer to keep it this way.

Passing bool is simple but I bit error prone - it creates implicit mapping between `ValidDisputeStatementKind` <-> `true` and `InvalidDisputeStatementKind` <-> `false`.

In the tests for example a lot of such values are created and it was a bit painful to set `true`/`false` for each call. In the current state you just pass the XXXDisputeStatementKind instance and don't care.

Does this make sense?

Reply:
I suggested a different approach previously, namely making use of [DisputeStatement](https://github.com/paritytech/polkadot/blob/36268ecf3e208e2179fdcf9b433ddf9909e39705/primitives/src/v2/mod.rs#L1215).

Reply:
Okay, I'll change this.